<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>promise的几个关键问题5</title>
</head>
<body>
<script>
    /*
    * 中断promise链
    *   1.当使用promise链式调用时，在中间中断，不再调用后边的回调函数
    *   2.方法：在失败状态的回调中返回一个pending状态的promise实例
    * */
    function sendAjax(url,data,method = 'GET') {
        return new Promise((resolve, reject) => {
            const xhr = new XMLHttpRequest();
            xhr.onreadystatechange = () => {
                if (xhr.readyState === 4){
                    if(xhr.status >= 200 && xhr.status <= 300) {
                        resolve(xhr.response)
                    }else {
                        reject('error')
                    }
                }
            };
            //整理参数
            let str = '';
            for (const key in data) {
                str += `${key}=${data[key]}&`
            }
            str = str.slice(0,-1)
            xhr.open(method ,url + '?' +str);
            // xhr.responseType = 'json'
            xhr.send();
        })
    }

    sendAjax('https://api.uixsj.cn/hitokoto/get',{type:'social'}).then(value => {
            console.log('第一次请求成功', value)
            return sendAjax('https://api.uixsj.cn/hitokoto/get2',{type:'social'})
        },err => {
            console.log('第一次请求失败', err)
            //中断promise链的方法：返回一个状态为pending状态的Promise实例
            return new Promise(()=>{})
        })
        .then(value => {
            console.log('第二次请求成功', value)
            return sendAjax('https://api.uixsj.cn/hitokoto/get',{type:'social'})
        },reason => {
            console.log('第二次请求失败', reason)
            return new Promise(()=>{})
        })
        .then(value => {
            console.log('第三次请求成功', value)
        },reason => {
            console.log('第三次请求失败', reason)
            return new Promise(()=>{})
        })
</script>
</body>
</html>
